### PR TITLE
Fix the height of the right column content

### PR DESCRIPTION
**Pull request by chrislomaxjones**
## What does this change?

This PR changes the way we layout the contents in the right right column - the `right` ad slot and the Most Viewed component.

The most viewed component can have variable height, as it depends on how much text is in the links on a given day. The right ad can have a variable height (up to 600px, depending on what is returned by the ad server) but is contained within a `1059px` region in which the advert is sticky.

This is the new revised layout:

image (1)

We fix the height of the container for both the right ad and and most viewed components to 1600px (this is approximately the height of the total components now, `(1059+530)px` where the `530px` is an estimate of the most viewed height) and arrange them as flex columns. Therefore, although the height of the most viewed can vary over time, this extra space will occupy the sticky region between the ad and the most viewed and not extend down into the right column (where we insert inline ads).

When an article is below 1600px in length, the most viewed component will not be rendered into the right column. It's worth noting this is different to the current implementation, where we attempt to determine if a `530px` Most Viewed could fit in the available space (thanks for pointing this out @domlander). I've decided to opt for simpler version for now, as I don't think this crops up in a lot of places and reduces the complexity of the measurement.

## Why?

Fixing the height of the right column content will allow Spacefinder to make more informed decisions about where it can safely float ads into the right column.

## Screenshots

Note there's not a lot of obvious visual difference between the new layout - except that the height at which the most viewed container will slightly change on most layouts. This can present issues with clashing with inline ads in the right column, as shown in the **before** below.
 
### Before:

https://user-images.githubusercontent.com/8000415/234847897-b84cf931-b112-49b9-ad63-1f08d34d704e.mov

### After:

https://user-images.githubusercontent.com/8000415/234848025-c1c013ff-09ef-48f3-b982-102ffb1c899c.mov

## What's not included

- I think the stickiness of the Most Viewed is broken when in ad free. This PR doesn't attempt to fix this, although it does shuffle some related props around.

- Spacefinder still isn't aware of the 1600px container height. Instead, it expects 1000px from the top of the first article paragraph. This seems to work most of the time, but we can separately make Spacefinder aware of this new height if we choose to do so.

### CODE REVIEW

File: dotcom-rendering/src/web/components/MostViewedRight.stories.tsx [hunk 1]
```diff
@@ -44,10 +44,7 @@ export const defaultStory = () => {
 						showTopBorder={false}
 						padSides={false}
 					>
-						<MostViewedRight
-							isAdFreeUser={false}
-							adBlockerDetected={false}
-						/>
+						<MostViewedRight />
```
Comment (chrislomaxjones) [inline dotcom-rendering/src/web/components/MostViewedRight.stories.tsx:47]: This is just shuffling some props around here.

File: dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx [hunk 2]
```diff
@@ -1,74 +1,43 @@
...
-	// Styling the data island root so it stretches to cover the full height available in the container.
-	// Requires us to subtract the height of its sibling in the container (Sticky top right ad slot).
-	const availableSpaceHeight = css`
-		height: calc(100% - ${rightAdvertHeight}px);
-	`;
+		const rightFurniture = document.querySelector<HTMLDivElement>(
+			'[data-component="right-column-content"]',
+		);
+		const height = rightFurniture?.getBoundingClientRect().height;
+		setHeightIsAvailable(height !== undefined && height >= 1600);
```
Comment (chrislomaxjones) [inline dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx:25]: This isn't super _Reacty_ but it seemed like the simplest way to get the height of an element outside of the island.

**Comment by chrislomaxjones** [comment]
I'll go ahead and approve this if there's no objections, but I'm reproducing it here to show that there is a slight change in position of the most viewed. This will be because it's now rendered at the **bottom** of the 1600px container, whereas before the **top** of it was rendered 1059px below the top of the right ad.

Screenshot 2023-04-27 at 10 21 57

**Review by domlander — COMMENTED** [review]
Most Viewed appears to be lazy loaded when the user reaches the bottom of the 1600px container, which means in can "jump" into the screen.

File: dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrislomaxjones) [reply dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx]: Nice idea - passing it down as a prop to avoid a cyclic dependency.

File: dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx [hunk 3]
```diff
@@ -1,74 +1,43 @@
...
-		? SHADY_PIE_HEIGHT
-		: TOP_RIGHT_AD_HEIGHT;
-
-	// Styling the data island root so it stretches to cover the full height available in the container.
-	// Requires us to subtract the height of its sibling in the container (Sticky top right ad slot).
-	const availableSpaceHeight = css`
-		height: calc(100% - ${rightAdvertHeight}px);
-	`;
+		const rightFurniture = document.querySelector<HTMLDivElement>(
+			'[data-component="right-column-content"]',
```
[the remark this answers is not in the document]
Comment (chrislomaxjones) [reply dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx]: As with `MAX_HEIGHT_PX`, will pass this down as a prop to avoid the cyclic dependency.

File: dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx [hunk 4]
```diff
@@ -1,74 +1,43 @@
...
+ * Wrapping MostViewedRight so we can determine whether or not there's enough vertical space in the container to render it
+ */
+export const MostViewedRightWrapper = ({ limitItems, renderAds }: Props) => {
 	const adBlockerDetected = useAdBlockInUse();
-	const bodyRef = useRef<HTMLDivElement>(null);
 	const [heightIsAvailable, setHeightIsAvailable] = useState<boolean>(false);
 
+	// We don't always show the most viewed component - it depends on the length of the article
+	// This effect determines whether to show the most viewed, depending on the computed height of the container
 	useEffect(() => {
```
[the remark this answers is not in the document]
Comment (chrislomaxjones) [reply dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx:27]: That's interesting thanks for sharing that. In my head this is how this component will behave, would be great if it matches what you think as well:

1) 1st Render. `heightIsAvailable` is `false`, therefore we return `null` and render nothing. The browser will perform a paint, and then go on to execute any effects.
2) Post 1st render we run the `useEffect`. We measure the container height, and then set `heightIsAvailable` to `true` if we determine there's enough space.
3) Since the state has changed we perform another re-render. If the height isn't available (i.e. no state change), nothing changes, so we don't re-render. If the height is available, we render the full most viewed component. The browser will then perform another paint.

Are we avoiding any flickering in this case, because we don't render the most viewed and then stop showing it in this case?

[the remark this answers is not in the document]
Comment (arelra) [reply dotcom-rendering/src/web/components/MostViewedRightWrapper.importable.tsx:27]: Ah yes exactly right. I missed `heightIsAvailable` is false on the first pass so it will only render the component once it has the height required so there should be no flicker.

**Review by arelra — APPROVED** [review]
Nice work!

**Comment by chrislomaxjones** [comment]
> Most Viewed appears to be lazy loaded when the user reaches the bottom of the 1600px container, which means in can "jump" into the screen.

Thanks for spotting this! 

I've snuck in a new prop `rootMargin` for islands that allows us to specify a much larger margin for the Most Viewed intersection observer, which should addresses this. From the commit message:

> Introduce a new prop rootMargin that can be added to islands that
> are deferred until visible. We then use this to add a significantly
> larger margin to the top of the most viewed, to account for the
> fact it is now loaded from the bottom of a 1600px container, instead
> of just under the 1059px right ad container.

File: dotcom-rendering/src/web/components/Island.tsx [hunk 5]
```diff
@@ -1,40 +1,61 @@
 import { Placeholder } from './Placeholder';
 
-interface HydrateProps {
-	deferUntil?: 'idle' | 'visible';
+type ClientOnlyProps = {
```
Comment (chrislomaxjones) [inline dotcom-rendering/src/web/components/Island.tsx:3]: This file is mostly a drive-by refactor on the props given to `Island` components - namely to separate out each of the different `deferUntil` properties and then separately to have different sets of props for when an Island is client only or not.

File: dotcom-rendering/src/web/components/Island.tsx [hunk 6]
```diff
@@ -1,40 +1,61 @@
...
+};
 
 /**
  * Props
  *
  * We use a union type here to support conditional typing. This means you
  * can only supply placeholderHeight if clientOnly is true.
  */
-type Props = HydrateProps | ClientOnlyProps | InteractionProps | HashProps;
+type Props =
```
Comment (jamesgorrie) [inline dotcom-rendering/src/web/components/Island.tsx:53]: I was trying to wrap my head around this, and had a couple of questions/thoughts:
* Why use the `param?: never` syntax in each type? Turns out we need this to destructure on those parameters in the `<Island />` component. Would these defaulting to `param?: never` in the `DefaultProps` and extending from there be clearer that that is why we do it?
* What is `ClientOnly` vs `NonClientOnly` Props? From what I could tell this was a conditional pair of properties defining the shape of `ClientOnly`. Would this make sense as `ClientOnlyProps` and set that conditional within that definition?

This was mainly trying to work out what was potting, so maybe not useful, but this was my conclusion in code, mainly based on readability:

```typescript
// We extend from these properties as we want to destructure on these
// properties on the `Island` component. Each state can override these individually
// otherwise the value will be `undefined`
type DefaultProps = {
	deferUntil?: never;
	clientOnly?: never;
	placeholderHeight?: never;
	rootMargin?: never;
	children: JSX.Element;
};

// This means you can only have `placeholderHeight` set when you have when
// clientOnly: true
type ClientOnlyProps =
	| {
			clientOnly: true;
			placeholderHeight?: number;
	  }
	| {
			clientOnly?: false;
			placeholderHeight?: never;
	  };

type IdleProps = DefaultProps &
	ClientOnlyProps & {
		deferUntil: 'idle';
		children: JSX.Element;
	};

type VisibleProps = DefaultProps &
	ClientOnlyProps & {
		deferUntil: 'visible';
		// see: https://developer.mozilla.org/en-us/docs/web/api/intersectionobserver/rootmargin
		rootMargin?: string;
		children: JSX.Element;
	};

type InteractionProps = DefaultProps & {
	deferUntil: 'interaction';
	children: JSX.Element;
};

type HashProps = DefaultProps & {
	deferUntil: 'hash';
	clientOnly: true;
	children: JSX.Element;
};

type Props =
	| DefaultProps
	| VisibleProps
	| IdleProps
	| InteractionProps
	| HashProps;
```


Hopefully it's not more confusing, but thought it worth sharing my workings out.

**Review by jamesgorrie — COMMENTED** [review]
Nothing blocking, but definitely wondering why we haven't used the `rootMargin` in any component? Is that for a future PR?

File: dotcom-rendering/src/web/components/RightColumnContent.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	isPaidContent: boolean;
+	renderAds: boolean;
+	shouldHideReaderRevenue: boolean;
+};
+
+/**
+ * The maximum height of the container that may contain
+ * a right ad slot and a most viewed component
+ */
+export const MAX_HEIGHT_PX = 1600;
```
[the remark this answers is not in the document]
Comment (chrislomaxjones) [reply dotcom-rendering/src/web/components/RightColumnContent.tsx:17]: Nice spot, this was imported somewhere before but isn't anymore. Thanks!

File: dotcom-rendering/src/web/components/Island.tsx (hunk 6 above)
Comment (chrislomaxjones) [reply dotcom-rendering/src/web/components/Island.tsx:53]: Thanks for sharing your thoughts on these types, I went a little bit back and forth on this change.

Anyway, my aim was to introduce an additional prop `rootMargin` that you can only supply when you include `deferUntil="visible"`, and cause a type error otherwise. But the way the current interfaces was structured meant that I had to split up the `HydrateProps` and `ClientOnlyProps` since they have `deferUntil?: 'visible' | 'idle'`.

Hence the discriminated union, where each of the possible types has a different discriminant `deferUntil` (the default case representing when you omit the `deferUntil` prop).

```typescript
type Props =
	| DefaultProps
	| VisibleProps
	| IdleProps
	| InteractionProps
	| HashProps;
```

Each type in the union then represents the full set of props permitted for that given defer option.

> What is ClientOnly vs NonClientOnly Props?

This is so, in both the `VisibleProps` and `IdleProps` case, you can have additional props that can only appear in two possible configurations (the client only configuration, where you specify `clientOnly={true}` and aren't allowed to specify a `placeholderHeight`, and the non client configuration where you may specify a `placeholderHeight` and may include `clientOnly={false}`, or omit entirely.

> Would these defaulting to param?: never in the DefaultProps and extending from there be clearer that that is why we do it?

I like the approach but I think having `DefaultProps` be one of the cases in the union and also set a bunch of types as `never` will prevent you from having islands of the form

```tsx
<Island clientOnly={true}>
```

I can add some TSDoc to these types, but also if this seems over-the-top I could just opt for simpler, more permissive types (aka less type safety, but maybe easier to understand?).

Comment (chrislomaxjones) [reply dotcom-rendering/src/web/components/Island.tsx:91]: I've actually nabbed your type for `ClientOnlyProps`, it's much neater. I've also added some TSDoc, let me know if you've got anymore thoughts.

File: dotcom-rendering/src/web/components/RightColumnContent.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+export const RightColumnContent = ({
```
[the remark this answers is not in the document]
Comment (chrislomaxjones) [reply dotcom-rendering/src/web/components/RightColumnContent.tsx:19]: Nice idea, gone for `MostViewedRightWithAd`.

**Review by domlander — APPROVED** [review]
